compiler/src/dotty/tools/dotc/interactive/Interactive.scala

```diff
@@ -165,38 +310,91 @@ object Interactive {
                    && (includeReferences || isDefinition(tree))
                    && treePredicate(tree))
                 buf += SourceTree(tree, source)
+              traverseChildren(tree)
+            case tree: untpd.Inlined =>
+              traverse(tree.call)
             case _ =>
+              traverseChildren(tree)
           }
-          traverseChildren(tree)
         }
       }).traverse(topTree)
     }
 
     buf.toList
   }
 
-  /** Check if `tree` matches `sym`.
-   *  This is the case if `sym` is the symbol of `tree` or, if `includeOverriden`
-   *  is true, if `sym` is overriden by `tree`.
-   */
-  def matchSymbol(tree: Tree, sym: Symbol, includeOverriden: Boolean)(implicit ctx: Context): Boolean =
-    (sym == tree.symbol) || (includeOverriden && tree.symbol.allOverriddenSymbols.contains(sym))
-
-
   /** The reverse path to the node that closest encloses position `pos`,
    *  or `Nil` if no such path exists. If a non-empty path is returned it starts with
    *  the tree closest enclosing `pos` and ends with an element of `trees`.
    */
   def pathTo(trees: List[SourceTree], pos: SourcePosition)(implicit ctx: Context): List[Tree] =
     trees.find(_.pos.contains(pos)) match {
-      case Some(tree) =>
-        // FIXME: We shouldn't need a cast. Change NavigateAST.pathTo to return a List of Tree?
-        val path = NavigateAST.pathTo(pos.pos, tree.tree, skipZeroExtent = true).asInstanceOf[List[untpd.Tree]]
+      case Some(tree) => pathTo(tree.tree, pos.pos)
+      case None => Nil
+    }
 
-        path.dropWhile(!_.hasType).asInstanceOf[List[tpd.Tree]]
-      case None =>
-        Nil
+  def pathTo(tree: Tree, pos: Position)(implicit ctx: Context): List[Tree] =
+    if (tree.pos.contains(pos)) {
+      // FIXME: We shouldn't need a cast. Change NavigateAST.pathTo to return a List of Tree?
+      val path = NavigateAST.pathTo(pos, tree, skipZeroExtent = true).asInstanceOf[List[untpd.Tree]]
+      path.dropWhile(!_.hasType).asInstanceOf[List[tpd.Tree]]
     }
+    else Nil
+
+  def contextOfStat(stats: List[Tree], stat: Tree, exprOwner: Symbol, ctx: Context): Context = stats match {
+    case Nil =>
+      ctx
+    case first :: _ if first eq stat =>
+      ctx.exprContext(stat, exprOwner)
+    case (imp: Import) :: rest =>
+      contextOfStat(rest, stat, exprOwner, ctx.importContext(imp, imp.symbol(ctx)))
+    case _ =>
+      ctx
+  }
+
+  def contextOfPath(path: List[Tree])(implicit ctx: Context): Context = path match {
```

Comment:
This looks very fragile. Since Typer already has to do a similar job for typedIdent to work, I was wondering if instead we could make a subclass of Typer that:
1. Skips trees outside of the current position
2. As soon as it reaches a leaf, find all completions and then exit.

Reply:
That's pretty much how scalac's PC works. It's a possibility, but would be a much larger rewrite compared to what we have. Other downsides: (1) it could be a lot slower, depending how much needs to be tpechecked. (2) we create ne symbols, which might confuse thiings. (3) Generally, more entanglement of normal compiler and presentation code.

Reply:
> (2) we create new symbols, which might confuse things.

But a ReTyper doesn't create symbols, does it? It also should be possible to not retypecheck anything and rely on the existing typechecked tree.

Reply:
That's true. Retypers don't create symbols. But I think messing with them will ultimately be more fragile. It comes down to whether we can coax a 3000 line module to do something quite different than what it was intended for, or whether we write a 40 line method that extracts the information directly. I was actually quite pleased how straightforward that was. True, there are some small differences, i.e. we ignore the "in supercall" distinctions. But I don't think these matter. And it's not clear that a retyper approach would not have similar differences.